ompi/mca/coll/ftagree/coll_ftagree_earlyreturning.c

```diff
@@ -0,0 +1,3367 @@
+                                         OMPI_NAME_PRINT(OMPI_PROC_MY_NAME),
+                                         ci->agreement_id.ERAID_FIELDS.contextid,
+                                         ci->agreement_id.ERAID_FIELDS.epoch,
+                                         ci->agreement_id.ERAID_FIELDS.agreementid,
+                                         r));
+                    break;
+                }
+            }
+            if( rl == (era_rank_item_t*)opal_list_get_end(&ci->gathered_info) ) {
+                OPAL_OUTPUT_VERBOSE((20, ompi_ftmpi_output_handle,
+                                     "%s ftagree:agreement (ERA) check_status for Agreement ID = (%d.%d).%d, some children have not contributed\n",
+                                     OMPI_NAME_PRINT(OMPI_PROC_MY_NAME),
+                                     ci->agreement_id.ERAID_FIELDS.contextid,
+                                     ci->agreement_id.ERAID_FIELDS.epoch,
+                                     ci->agreement_id.ERAID_FIELDS.agreementid));
+                /* We are still waiting for a message from at least a child. Let's wait */
+                return;
+            }
+        }
+
+        /* Left that loop? We're good to decide locally */
+        era_update_return_value(ci, -1, NULL);
+
+        if( ci->comm->c_my_rank == (r = era_parent(ci)) ) {
+            OPAL_OUTPUT_VERBOSE((20, ompi_ftmpi_output_handle,
+                                 "%s ftagree:agreement (ERA) check_status for Agreement ID = (%d.%d).%d, all children of root have contributed\n",
+                                 OMPI_NAME_PRINT(OMPI_PROC_MY_NAME),
+                                 ci->agreement_id.ERAID_FIELDS.contextid,
+                                 ci->agreement_id.ERAID_FIELDS.epoch,
+                                 ci->agreement_id.ERAID_FIELDS.agreementid));
+
+            /* I'm root. I have to decide now. */
+            era_decide(ci->current_value, ci);
+        } else {
+            OPAL_OUTPUT_VERBOSE((20, ompi_ftmpi_output_handle,
+                                 "%s ftagree:agreement (ERA) check_status for Agreement ID = (%d.%d).%d, all children of non-root have contributed\n",
+                                 OMPI_NAME_PRINT(OMPI_PROC_MY_NAME),
+                                 ci->agreement_id.ERAID_FIELDS.contextid,
+                                 ci->agreement_id.ERAID_FIELDS.epoch,
+                                 ci->agreement_id.ERAID_FIELDS.agreementid));
+
+            /* Let's forward up and wait for the DOWN messages */
+            ci->waiting_down_from = r;
+            send_msg(ci->comm, r, NULL, ci->agreement_id, MSG_UP, ci->current_value,
+                     ci->nb_acked, ci->acked);
+            ci->status = BROADCASTING;
+        }
+        return;
+    }
+
+    opal_output(0, "%s: %s (%s:%d) -- Need to implement function for that case (status = %d)\n",
```

Comment:
This is a fatal error, I'm not sure the message is right, and probably OMPI uses another approach than assert(0) to signal internal errors?

Reply:
What is 'that case' by the way?

Reply:
That would be if we call check_status while in BROADCASTING or COMPLETED states. That would be wrong, internally: we should only call check_status if the change may make us decide. It doesn't happen (I don't think you ever saw it in your tests? I haven't), but this test was here to catch code logic problems, during development.

Reply:
SO it's an assert of sort.